spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
+    IF minTrustedHeight > heightToTrust \* outside of the trusting period
+        \/ ~(signedHdr[2] \subseteq (DOMAIN signedHdr[1].VP)) \* signed by other nodes
+    THEN FALSE
+    ELSE
+      LET TP == BC!PowerOfSet(trustedHdr.NextVP, DOMAIN trustedHdr.NextVP)
+          SP == BC!PowerOfSet(trustedHdr.NextVP,
+            signedHdr[2] \intersect DOMAIN trustedHdr.NextVP)
+      IN
+      IF heightToVerify = heightToTrust + 1
+          \* the special case of adjacent heights: check 2/3s
+      THEN (3 * SP > 2 * TP)
+      ELSE \* the general case: check 1/3 between the headers and make sure there are 2/3 votes  
+        LET TPV == BC!PowerOfSet(signedHdr[1].VP, DOMAIN signedHdr[1].VP)
+            SPV == BC!PowerOfSet(signedHdr[1].VP, signedHdr[2])
+        IN
+        (3 * SP > TP) /\ (3 * SPV > 2 * TPV)
+
+(* Make one step of bisection, roughly one stack frame of Bisection in the English spec *)
+OneStepOfBisection(storedHdrs) ==
+    LET topReq == Head(requestStack)
+        lh == topReq.startHeight
+        rh == topReq.endHeight
+        lhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = lh
+        rhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = rh
+    IN
+    \* pass only the header lhdr[1] and signed header rhdr
+    IF CheckSupport(lh, rh, lhdr[1], rhdr)
+        (* The header can be trusted, pop the request and return true *)
+    THEN <<TRUE, Tail(requestStack)>>
+    ELSE IF lh + 1 = rh \* sequential verification
+        THEN (*
+             Sequential verification tells us that the header cannot be trusted:
+             If the search request was scheduled as a left branch, then
+             (1) pop the top request (the left branch), and
+             (2) pop the second top request, which is the request for the right branch.
+             Otherwise, pop only the top request.
+             This is the optimization that is introduced in the English spec.
+             One can replace it e.g. with a parallel search.
+             *)
+            LET howManyToPop == IF topReq.isLeft THEN 2 ELSE 1 IN  
+            <<FALSE, SubSeq(requestStack, 1 + howManyToPop, Len(requestStack))>>
+        ELSE (*
+             Dichotomy: schedule search requests for the left and right branches
+             (and pop the top element off the stack).
+             In contrast to the English spec, these requests are not processed immediately,
+             but one-by-one in a depth-first order (mind the optimization above).
+             *)
+            LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
+                leftReq ==  [isLeft |-> TRUE, startHeight |-> lh, endHeight |-> (lh + rh) \div 2]
+            IN
+            <<TRUE, <<leftReq, rightReq>> \o Tail(requestStack)>>
+
+OnResponseHeader ==
+  /\ state = "working"
+  /\ inEvent.type = "responseHeader"
+  /\ storedHeaders' = storedHeaders \union { inEvent.hdr }
```

Comment:
Should we store all headers or only valid one?

Reply:
All, as we can use them later as a proof.